pkg/controller/worker/machine_dependencies.go

```diff
@@ -1,25 +1,181 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
 
 package worker
 
-import "context"
+import (
+	"context"
+	"fmt"
 
-func (w *workerDelegate) DeployMachineDependencies(_ context.Context) error {
-	return nil
+	api "github.com/gardener/gardener-extension-provider-openstack/pkg/apis/openstack"
+	"github.com/gardener/gardener-extension-provider-openstack/pkg/apis/openstack/helper"
+	osclient "github.com/gardener/gardener-extension-provider-openstack/pkg/openstack/client"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+)
+
+func (w *workerDelegate) DeployMachineDependencies(ctx context.Context) error {
+	computeClient, err := w.openstackClient.Compute()
+	if err != nil {
+		return err
+	}
+
+	workerStatus, err := w.decodeWorkerProviderStatus()
+	if err != nil {
+		return err
+	}
+
+	serverGroupDepSet, err := w.reconcileServerGroups(computeClient, workerStatus.DeepCopy())
+	return w.updateMachineDependenciesStatus(ctx, workerStatus, serverGroupDepSet.extract(), err)
+}
+
+func (w *workerDelegate) reconcileServerGroups(computeClient osclient.Compute, workerStatus *api.WorkerStatus) (serverGroupDependencySet, error) {
+	serverGroupDepSet := newServerGroupDependencySet(workerStatus.ServerGroupDependencies)
+	for _, pool := range w.worker.Spec.Pools {
+		serverGroupDependencyStatus, err := w.reconcilePoolServerGroup(computeClient, pool, serverGroupDepSet)
+		if err != nil {
+			return serverGroupDepSet, fmt.Errorf("reconciling server groups failed for pool %q: %w", pool.Name, err)
+		}
+		serverGroupDepSet.upsert(serverGroupDependencyStatus)
+	}
+	return serverGroupDepSet, nil
+}
+
+func (w *workerDelegate) reconcilePoolServerGroup(computeClient osclient.Compute, pool extensionsv1alpha1.WorkerPool, set serverGroupDependencySet) (*api.ServerGroupDependency, error) {
+	poolProviderConfig, err := helper.WorkerConfigFromRawExtension(pool.ProviderConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	if poolProviderConfig == nil || poolProviderConfig.ServerGroupConfig == nil || poolProviderConfig.ServerGroupConfig.Policy == "" {
+		return nil, nil
+	}
+
+	poolDep := set.getByPoolName(pool.Name)
+	if poolDep != nil {
+		serverGroup, err := computeClient.GetServerGroup(poolDep.ID)
+		if err != nil && !osclient.IsNotFoundError(err) {
```

Comment:
What's the expected behaviour when a "not found" error is returned? At the moment, it's silently ignored, right? Just to double-check that one.

Reply:
Yes, we silently ignore it. We will create a new server group and overwrite the old serverGroupID with the new in our worker's dependencies status. 

Later in the worker reconciliation we will garbage collect all serverGroups that match our "name-tag" `clusterName-workerName` but have no reference in our dependencies status.

Reply:
OK, thanks